src/main/scala/beam/physsim/PickUpDropOffCollector.scala

```diff
@@ -0,0 +1,277 @@
+
+import beam.agentsim.agents.vehicles.BeamVehicleType
+import beam.agentsim.events.PathTraversalEvent
+import beam.router.Modes
+import beam.sim.config.BeamConfig
+import com.typesafe.scalalogging.LazyLogging
+import org.matsim.api.core.v01.Id
+import org.matsim.api.core.v01.events.{Event, PersonEntersVehicleEvent, PersonLeavesVehicleEvent}
+import org.matsim.core.controler.events.{IterationEndsEvent, IterationStartsEvent}
+import org.matsim.core.controler.listener.{IterationEndsListener, IterationStartsListener}
+import org.matsim.core.events.handler.BasicEventHandler
+import org.matsim.vehicles.Vehicle
+
+import scala.annotation.tailrec
+import scala.collection.mutable
+
+class PickUpDropOffCollector(vehicleTypes: Map[Id[BeamVehicleType], BeamVehicleType])
+    extends BasicEventHandler
+    with IterationStartsListener
+    with LazyLogging {
+
+  private val vehicleToPickUpsDropOffs: mutable.HashMap[Id[Vehicle], VehiclePickUpsDropOffs] = mutable.HashMap.empty
+  private val personsPickUps: mutable.ListBuffer[PersonPickUp] = mutable.ListBuffer.empty
+  private val personsDropOffs: mutable.ListBuffer[PersonDropOff] = mutable.ListBuffer.empty
+
+  private val vehiclesToIgnore: mutable.HashSet[Id[Vehicle]] = mutable.HashSet.empty
+  private val vehicleTypeToCAV: mutable.HashMap[String, Boolean] = mutable.HashMap.empty
+
+  private def isRideHail(vehicleId: Id[Vehicle]): Boolean = vehicleId.toString.startsWith("rideHailVehicle")
+
+  private var maybePickUpDropOffHolder: Option[PickUpDropOffHolder] = None
+
+  private def isCAV(vehicleType: String): Boolean = {
+    vehicleTypeToCAV.get(vehicleType) match {
+      case None =>
+        val beamVehicleTypeId = Id.create(vehicleType, classOf[BeamVehicleType])
+        val isCaccEnabled = vehicleTypes.get(beamVehicleTypeId) match {
+          case Some(beamVehicleType) => beamVehicleType.isCaccEnabled
+          case _                     => false
+        }
+
+        vehicleTypeToCAV.put(vehicleType, isCaccEnabled)
+        isCaccEnabled
+
+      case Some(isCaccEnabled) => isCaccEnabled
+    }
+  }
+
+  private def isRideHailOrCAV(pte: PathTraversalEvent): Boolean = {
+    val mode = pte.mode
+    mode match {
+      case Modes.BeamMode.CAR if isRideHail(pte.vehicleId) || isCAV(pte.vehicleType) => true
+      case Modes.BeamMode.CAV                                                        => true
+      case Modes.BeamMode.RIDE_HAIL                                                  => true
+      case Modes.BeamMode.RIDE_HAIL_POOLED                                           => true
+      case Modes.BeamMode.RIDE_HAIL_TRANSIT                                          => true
+      case _                                                                         => false
+    }
+  }
+
+  override def handleEvent(event: Event): Unit = {
+    def vehicleNotIgnored(vehicleId: Id[Vehicle]): Boolean = {
+      val isIgnored = vehiclesToIgnore.contains(vehicleId)
+      !isIgnored
+    }
+
+    event match {
```

Comment:
Repositioning of ride-hail fleet (controlled by `beam.agentsim.agents.rideHail.repositioningManager.name`) will cause them to be driving around the map which will create `PathTraversalEvent` every time when ride-hail vehicle repositions. Shall it be specifically handled because there will no no customer in that vehicle?

Reply:
I do not think that empty RH vehicles should be handled here, at least for now.